data/nicotine.1
.\" SPDX-FileCopyrightText: 2006-2025 Nicotine+ Contributors
.\" SPDX-License-Identifier: GPL-3.0-or-later

.TH NICOTINE+ 1

.SH NAME
nicotine - graphical client for the Soulseek peer-to-peer network

.SH SYNOPSIS
.B nicotine
[options]

.SH DESCRIPTION
.BI Nicotine+
is a graphical application that allows users to interact with the
Soulseek peer-to-peer network.

.SH OPTIONS
.TP
.BI \-c " <file>" "\fR,\fP \-\^\-config=" <file>
Use <file> as configuration file. If <file> does not exists a new empty
configuration file will be created.
.TP
.BI \-u " <dir>" "\fR,\fP \-\^\-user-data=" <dir>
Use <dir> as directory for user data and plugins. If <dir> does not
exists a new empty directory will be created.
.TP
.B \-h, \-\^\-help
Show command help and exit.
.TP
.B \-s, \-\^\-hidden
Start the program without showing window.
.TP
.BI \-b " <ip>" "\fR,\fP \-\^\-bindip=" <ip>
Bind sockets to the given <ip> (useful for VPN).
.TP
.BI \-l " <port>" "\fR,\fP \-\^\-port=" <port>
Listen on the given port. Overrides the port range configuration.
.TP
.B \-r, \-\^\-rescan
Rescan shared files.
.TP
.B \-n, \-\^\-headless
Start the program in headless mode (no GUI).
.TP
.B \-v, \-\^\-version
Show version number and exit.

.SH EXIT STATUS
The regular exit status of the program is 0.
.br
If required dependencies are not present, the exit status is 1.
.br
If the given list of arguments cannot be parsed, the exit status is 2.

.SH FILES
.TP
.I ~/.config/nicotine/config
The configuration file.
.TP
.I ~/.local/share/nicotine/
Other personal files such as downloads, shares and logs.

.SH ENVIRONMENT
The following environment variables are available when running the program.
.TP
.IP NICOTINE_GTK_VERSION
The major GTK version to use, if installed. Possible values are 4 or 3.
.TP
.IP NICOTINE_LIBADWAITA
Whether to enable libadwaita (GNOME theme), if installed. Possible values
are 1 or 0.

.SH AUTHOR
Nicotine+ Team <https://nicotine-plus.org/>


NEWS
News
====

    As per GCS § 6.7, this file contains a list of user-visible noteworthy
    changes. Note that this is not the same as a changelog.

Version 2.0.0 (14 July 2020)
-----------------------------

Changes
 * Ported from Python 2 to Python 3
 * Ported from GTK2 to GTK3 (PyGTK to PyGObject)
 * Support for HiDPI displays
 * Search results and transfers can now be grouped by folder
 * Support for transfers larger than 2 GB in size
 * Transfers and search results now support drag-select
 * Performance improvements in downloads, uploads and search views
 * Special characters (e.g. -, ') are now removed from search terms by default, to receive more search results.
   This behavior can be toggled in Settings -> Misc -> Searches.
 * Excluding search results by placing a - sign in front of a word now works properly
 * Search filters now check the directory path
 * Column widths are now remembered between sessions
 * Added option to open previous tab on startup
 * Added option to hide buddy list
 * Custom messages can now be sent to leechers in Settings -> Misc -> Plugins -> Leech detector
 * Plugins are now bundled with Nicotine+ installations by default
 * Nicotine+ now follows the XDG Base Directory Specification
 * Replaced deprecated dependencies with maintained ones
 * Added unit and DEP-8 continuous integration testing
 * Minor UI cleanups
 * General code cleanups, removed dead code
 * Replaced non-free sound effects

Bugs closed in GitHub
 * Columns Position Not Being Maintained (#8)
 * Add "Group by folder" option to search results (#17)
 * Downloads tab hanging when adding a lot of files (#34)
 * NTFS support on linux (#49)
 * Show network drives when adding a shared directory. (#52)
 * send to player does not work. (#53)
 * CPU usage spikes and remains high after period of usage (#54)
 * Segfault When Getting User Info (#57)
 * Segmentation fault on Ubuntu Gnome 17.04 (#60)
 * filenames with ? in them get stuck on uploads list (#61)
 * Nicotine+ Windows 8.1 (64-bit) mutagen attempts to handle non-video files (#62)
 * Nicotine+ 1.4.1, windows 8.1 (64-bit) errors when using UPNP (#63)
 * Nicotine+ 1.4.1, windows 8.1 (64-bit) Spurious error messages (#64)
 * Nicotine + 1.4.1, windows 8.1 (64-bit) buttons not working (#65)
 * Downloads directory is not shared (#66)
 * Can't share directories (#68)
 * Question: Is Development Dead? (#73)
 * select ValueError: filedescriptor out of range in select() (#77)
 * blurry tray icon in kde plasma (#81)
 * Problems sharing files (#83)
 * Choosing "Download containing folder(s)" from search results does nothing (#84)
 * Uploads not working (#85)
 * UI very condensed on high-dpi linux. (#88)
 * Wishlist returns empty results for foreign characters (#89)
 * New Commits - Is Development Back? (#90)
 * Filter doesn't include directory path (#91)
 * XDG Base Directory Support (#94)
 * Port to python3 (#99)
 * Nicotine+ 1.4.2, Debian 9 (64-bit) Downloading file size >2GB appears as negative numbers, files near 4GB download 0 byte. (#100)
 * Nicotine+ 1.4.1 don't handle invalid characters in Windows (#101)
 * Random crash on Raspbian (#103)
 * Bitrate not shown for most music in search results (#104)
 * Nicotine+ 1.4.2, Debian 9 (64 bit) : Can't get shared files + current downloads disappeared : since the last but one update, from branch master (#107)
 * Website is badly out of date (#109)
 * images seem to be integrated from the launch directory if they have special names (#113)
 * Not working on Ubuntu 20.04 Focal Fossa (#115)
 * Please update Nicotine to work on the latest Ubuntu (20.04) (#123)
 * Compiled 'Master Branch' - Nicotine is Black Blank Screen? (#140)
 * Question: 1.4.3 - Columns Hiding? (#143)
 * info user correct extra typo (#144)
 * select user transfert does not select anything (#145)
 * clicking hyperlinks does not open browser (#146)
 * left click does not work on users nickname in rooms (#147)
 * Interest tab : text zone too small (#148)
 * request : adding file chooser preview widget in info user picture setting (#149)
 * menu separator does not follow gtk+ rules (#151)
 * 1.4.3 Linux - Hidding Tabs - Always Opens Now Under Buddy List (#154)
 * strace shows weird file access syscalls (#155)
 * (world) flags missing at startup / and buddy list (#161)
 * setup.py: DistutilsFileError (#164)
 * warnings causes by userlist resizing columns (#165)
 * Question: No more charsets selection ? (#180)
 * my gtk3 theme gives checkbuttons looks bigger (#181)
 * Question - Bug? - Log Window Issue (#186)
 * wait a minute, only spellchecker is missing ? (#190)
 * userlist for myself does not display files number (#192)
 * AttributeError in changecolour(): PrivateChat object has no attribute 'tag_log' (#194)
 * Add support for >2GB downloads (#201)
 * IndexError at start on Debian Buster (#202)
 * Speed up program startup times (#215)
 * custom tray icons not respected (#239)
 * Request: Modes Tab Placement? (#242)
 * text in log aera in chat rooms lag to display from entry (#253)
 * /now playing does not work after nic+ restart (#255)
 * add grouping by path (#269)
 * on kde LMB on tray icon brings menu, not app (#270)
 * lower on an int? (#278)
 * right-clicking file that user 2 downloads points to user 1 (#297)
 * Private Chat tab does not get notified on receiving a message (#299)
 * RMB doesn't select what's underneath it (#300)
 * unable to download to created folder (#301)
 * status never reach 100% becasue of filtered files (#302)
 * twice downloaded same folder, aborted duplicate files, remove aborted does not remove (#305)
 * downloading folder from user browse doesn't work (#311)
 * cannot connect (#312)
 * In download page, pressing Delete key removes 2 files instead of 1 (#314)
 * invalid path (#318)
 * Distrib message type 93 unknown (#322)
 * Connection issues after search (#329)
 * Window decorator close button doesn't work (#330)
 * Question: group by folders vs group by users (#335)
 * #312 continued, cannot connect (#336)

nicotine.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
#
# COPYRIGHT (c) 2016 Michael Labouebe <gfarmerfr@free.fr>
# COPYRIGHT (c) 2008-2011 Quinox <quinox@users.sf.net>
# COPYRIGHT (c) 2006-2008 eL_vErDe <gandalf@le-vert.net>
# COPYRIGHT (C) 2006-2009 Daelstorm <daelstorm@gmail.com>
#
# GNU GENERAL PUBLIC LICENSE
#    Version 3, 29 June 2007
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

"""
Nicotine+ Launcher.
"""

import os
import platform
import sys
from pynicotine.logfacility import log
from pynicotine.utils import ApplyTranslation

# Setting gettext and locale
ApplyTranslation()

# Detect if we're running on Windows
win32 = platform.system().startswith("Win")


def checkenv():

    # Require Python 2.7
    try:
        assert sys.version_info[:2] == (2, 7), '.'.join(
            map(str, sys.version_info[:3])
        )
    except AssertionError as e:
        return _("""You're using an unsupported version of Python (%s).
You should install Python 2.7.X.""") % (e)

    # Require GTK+ 2.24
    try:
        import gtk
    except Exception as e:
        return _("Cannot find GTK+ 2.24.X, please install it.")
    else:
        try:
            assert gtk.gtk_version[:2] == (2, 24), '.'.join(
                map(str, gtk.gtk_version[:3])
            )
        except AssertionError as e:
            return _("""You're using an unsupported version of GTK (%s).
You should install GTK 2.24.X.""") % (e)

    # Require PyGTK 2.24
    try:
        import pygtk
    except Exception as e:
        return _("Cannot find PyGTK 2.24.X, please install it.")
    else:
        try:
            assert gtk.pygtk_version[:2] == (2, 24), '.'.join(
                map(str, gtk.pygtk_version[:3])
            )
        except AssertionError as e:
            return _("""You're using an unsupported version of PyGTK (%s).
You should install PyGTK 2.24.X.""") % (e)

    # Require pynicotine
    try:
        import pynicotine
    except ImportError, e:
        return _("""Can not find Nicotine+ modules.
Perhaps they're installed in a directory which is not
in an interpreter's module search path.
(there could be a version mismatch between
what version of python was used to build the Nicotine
binary package and what you try to run Nicotine+ with).""")

    # Check for optionnal GeoIP
    try:
        import GeoIP
    except ImportError:
        try:
            import _GeoIP
        except ImportError:
            msg = _("""Nicotine+ supports a country code blocker.
This requires a (GPL'ed) library called GeoIP. You can find it here:
C library:       http://www.maxmind.com/app/c
Python bindings: http://www.maxmind.com/app/python
(the python bindings require the C library)""")
            log.addwarning(msg)

    # Windows specific checks
    if win32:

        # dbhash might be a good choice
        try:
            import dbhash
        except:
            log.addwarning(
                _("Warning: the Berkeley DB module, dbhash, " +
                  "could not be loaded."))

        # win32file is used to handle hidden files
        try:
            import win32file
        except:
            msg = _("""Nicotine+ supports hidden files attributes on windows.
This requires a python module called pypiwin32. You can find it here:
https://pypi.python.org/pypi/pypiwin32""")
            log.addwarning(msg)

    return None


def version():
    try:
        import pynicotine.utils
        print _("Nicotine+ version %s" % pynicotine.utils.version)
    except ImportError, error:
        print _("Cannot find the pynicotine.utils module.")


def usage():
    print _("""Nicotine+ is a Soulseek client.
Usage: nicotine [OPTION]...
  -c file, --config=file      Use non-default configuration file
  -p dir,  --plugins=dir      Use non-default directory for plugins
  -t,      --enable-trayicon  Enable the tray icon
  -d,      --disable-trayicon Disable the tray icon
  -h,      --help             Show help and exit
  -s,      --hidden           Start the program hidden so only the tray icon is shown
  -b ip,   --bindip=ip        Bind sockets to the given IP (useful for VPN)
  -v,      --version          Display version and exit""")


def renameprocess(newname, debug=False):

    errors = []

    # Renaming ourselves for ps et al.
    try:
        import procname
        procname.setprocname(newname)
    except:
        errors.append("Failed procname module")

    # Renaming ourselves for pkill et al.
    try:
        import ctypes
        # GNU/Linux style
        libc = ctypes.CDLL('libc.so.6')
        libc.prctl(15, newname, 0, 0, 0)
    except:
        errors.append("Failed GNU/Linux style")

    try:
        import dl
        # FreeBSD style
        libc = dl.open('/lib/libc.so.6')
        libc.call('setproctitle', newname + '\0')
        renamed = True
    except:
        errors.append("Failed FreeBSD style")

    if debug and errors:
        msg = [_("Errors occured while trying to change process name:")]
        for i in errors:
            msg.append("%s" % (i,))
        log.addwarning('\n'.join(msg))


def run():

    renameprocess('nicotine')

    import getopt
    import os.path
    try:
        opts, args = getopt.getopt(sys.argv[1:],
                                   "hc:p:tdvswb:",
                                   [
                                    "help",
                                    "config=",
                                    "plugins=",
                                    "profile",
                                    "enable-trayicon",
                                    "disable-trayicon",
                                    "version",
                                    "hidden",
                                    "bindip="
                                   ]
                                   )
    except getopt.GetoptError:
        # print help information and exit
        usage()
        sys.exit(2)

    if win32:
        try:
            mydir = os.path.join(os.environ['APPDATA'], 'nicotine')
        except KeyError:
            mydir, x = os.path.split(sys.argv[0])
        config = os.path.join(mydir, "config", "config")
        plugins = os.path.join(mydir, "plugins")
    else:
        config = os.path.join(os.path.expanduser("~"), '.nicotine', 'config')
        plugins = os.path.join(os.path.expanduser("~"), '.nicotine', 'plugins')

    profile = 0
    trayicon = 1
    hidden = False
    bindip = None

    for o, a in opts:
        if o in ("-h", "--help"):
            usage()
            sys.exit()
        if o in ("-c", "--config"):
            config = a
        if o in ("-p", "--plugins"):
            plugins = a
        if o in ("-b", "--bindip"):
            bindip = a
        if o == "--profile":
            profile = 1
        if o in ("-t", "--enable-trayicon"):
            trayicon = 1
        if o in ("-d", "--disable-trayicon"):
            trayicon = 0
        if o in ('-s', '--hidden'):
            hidden = True
        if o in ("-v", "--version"):
            version()
            sys.exit()

    result = checkenv()

    if result is None:
        from pynicotine.gtkgui import frame

        app = frame.MainApp(config, plugins, trayicon, hidden, bindip)

        if profile:
            import hotshot
            logfile = os.path.expanduser(config) + ".profile"
            profiler = hotshot.Profile(logfile)
            log.add(_("Starting using the profiler (saving log to %s)") %
                    logfile)
            profiler.runcall(app.MainLoop)
        else:
            app.MainLoop()
    else:
        print result

if __name__ == '__main__':
    try:
        run()
    except SystemExit:
        raise
    except Exception:
        import traceback
        traceback.print_exc()
